### ISSUE TITLE
Add ability to use password for DataSource tunnel

### ISSUE THREAD

**Issue by flinkgutt** [lucastheisen/jsch-extension#1]
It would be useful to have the ability to set a password instead of using ssh keys, while I understand it's not exactly "ideal" it would be useful (for me at least).

**Comment by meshcow** [comment]
strong vouch!

At the moment the only easy way I've found is to use the following
```java
class MySessFactory extends DefaultSessionFactory {
        private String passwd;

        public MySessFactory(String username, String hostname, Integer port, String passwd) {
            super(username, hostname, port);
            this.passwd = passwd;
        }

        @Override
        public Session newSession() throws JSchException {
            Session session= super.newSession();
            session.setPassword(passwd);
            return session;
        }

        @Override
        public SessionFactoryBuilder newSessionFactoryBuilder() {
            return new SessionFactoryBuilder( null, getUsername(), getHostname(), getPort(), getProxy(), null ) {
                @Override
                public SessionFactory build() {
                    return new MySessFactory( username, hostname, port, passwd );
                }
            };
        }
    }
 ```
instead of DefaultSessionFactory, which is a kind of a dirty hack (jsch & config fields are private in parent class ...).
For sure one might implement proper SessionFactory, but adding a password injection to existing DefaultSessionFactory seems a lot easier

**Comment by lucastheisen** [comment]
@meshcow  or @flinkgutt, I hear ya!  I have added support and pushed the latest fix.  I don't currently have time to test.  Can you build/test master to verify that it satisfies your requirements?  (And if you're feeling really generous, a unit test pull request would be nice)...  Once you verify, i will be happy to build a new release for you.

**Comment by meshcow** [comment]
Voila, requires a config property "scp.out.test.password"

```java
package com.pastdev.jsch;

import com.pastdev.jsch.command.CommandRunner;
import org.junit.Assume;
import org.junit.BeforeClass;
import org.junit.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import java.io.IOException;
import java.io.InputStream;
import java.util.Properties;
import static org.junit.Assert.*;

public class PasswordConnectivityTest {
    private static Logger logger = LoggerFactory.getLogger(PasswordConnectivityTest.class);
    private static String hostname;
    private static int port;
    private static String username;
    private static String correctPassword;
    private static String incorrectPassword;

    @BeforeClass
    public static void initializeClass() {
        InputStream inputStream = null;
        Properties properties = null;
        try {
            inputStream = ClassLoader.getSystemResourceAsStream("configuration.properties");
            Assume.assumeNotNull(inputStream);
            properties = new Properties();
            properties.load(inputStream);
        } catch (IOException e) {
            logger.warn("cant find properties file (tests will be skipped): {}", e.getMessage());
            Assume.assumeNoException(e);
        } finally {
            IOUtils.closeAndLogException(inputStream);
        }
        username = properties.getProperty("scp.out.test.username");
        hostname = properties.getProperty("scp.out.test.host");
        correctPassword = properties.getProperty("scp.out.test.password");
        assertNotNull(correctPassword);
        incorrectPassword = correctPassword + ".";
        port = Integer.parseInt(properties.getProperty("scp.out.test.port"));
    }

    private SessionFactory getSessionFactory(String password) throws IOException {
        final DefaultSessionFactory defaultSessionFactory = new DefaultSessionFactory(username, hostname, port);
        defaultSessionFactory.setConfig("PreferredAuthentications", "keyboard-interactive");  // to exclude publickey auth
        defaultSessionFactory.setPassword(password);
        return defaultSessionFactory;
    }

    private void testConnectionWithPassword(String password) throws Exception {
        CommandRunner commandRunner = null;
        try {
            commandRunner = new CommandRunner(getSessionFactory(password));
            final String expected = "hello";
            CommandRunner.ExecuteResult result = commandRunner.execute("echo " + expected);
            assertEquals(0, result.getExitCode());
            assertEquals(expected + "\n", result.getStdout());
            assertEquals("", result.getStderr());
        } finally {
            IOUtils.closeAndLogException(commandRunner);
        }
    }

    @Test
    public void testConnectionWithCorrectPassword() {
        try {
            testConnectionWithPassword(correctPassword);
        } catch (Exception e) {
            fail(e.getMessage());
        }
    }

    @Test
    public void testConnectionWithIncorrectPassword() {
        try {
            testConnectionWithPassword(incorrectPassword);
        } catch (Exception e) {
            assertEquals(e.getClass(), com.jcraft.jsch.JSchException.class);
            assertEquals(e.getMessage(), "Auth cancel");
        }
    }
}
```

So I've tested and can confirm it works. 
I'm using https://github.com/lucastheisen/jsch-nio/, can I expect it will be also updated, please?

**Comment by lucastheisen** [comment]
@meshcow, I have released [jsch-extension-0.1.5](https://github.com/lucastheisen/jsch-extension/tree/jsch-extension-0.1.5), and [jsch-nio-0.1.8](https://github.com/lucastheisen/jsch-nio/tree/jsch-nio-0.1.8)

However, I think it is worth noting that your unit test did not work for me as `password` is _NOT_ the same as `keyboard-interactive`.  The underlying JSch library will work _in most cases_ when you `setPassword` and use `keyboard-interactive`, but didn't work in my environment (cygwin sshd on windows).  But switching the `PreferredAuthentications` to `password` allowed the unit test to pass.  I also added a `setUserInfo` to the `DefaultSessionFactory` in cause you really do need `keyboard-interactive`.  

Anyway, the releases should be ready for you soon.

**Comment by meshcow** [comment]
In my case it didn't work with `password` because my remote targets (osx & ubuntu) both had `PasswordAuthentication no` in sshd_config. The `keyboard-interctive` allowed me to use password as both machines had `ChallengeResponseAuthentication yes` in the conf.
So the correct way to run it actually is to use
```java
defaultSessionFactory.setConfig("PreferredAuthentications", "keyboard-interactive,password");
```
Sorry, it was my fault; should have specified both auth methods there.
